pyke/prf_photometry.py

```diff
@@ -0,0 +1,407 @@
+from oktopus.posterior import PoissonPosterior
+from .utils import channel_to_module_output, plot_image
+
+if sys.version_info[0] == 2:
+    from inspect import getargspec
+    def _get_number_of_arguments(func):
+        return len(getargspec(func).args)
+
+else:
+    from inspect import signature
+    def _get_number_of_arguments(func):
+        return len(signature(func).parameters)
+
+
+__all__ = ['PRFPhotometry', 'SceneModel', 'KeplerPRF', 'SimpleKeplerPRF', 'get_initial_guesses']
+
+
+class PRFPhotometry(object):
+    """
+    This class performs PRF Photometry on TPF-like files.
+
+    Attributes
+    ----------
+    scene_model : instance of SceneModel
+        Model which will be fit to the data
+    priors : instance of oktopus.JointPrior
+        Priors on the parameters that will be estimated
+    loss_function : subclass of oktopus.LossFunction
+        Noise distribution associated with each random measurement
+    """
+
+    def __init__(self, scene_model, prior, loss_function=PoissonPosterior, **kwargs):
+        self.scene_model = scene_model
+        self.prior = prior
+        self.loss_function = loss_function
+        self.loss_kwargs = kwargs
+        self.opt_params = np.array([])
+        self.residuals = np.array([])
+        self.loss_value = np.array([])
+        self.uncertainties = np.array([])
+
+    def fit(self, tpf_flux, x0=None, cadences='all', method='powell',
+            **kwargs):
+        """
+        Fits the scene model to the given data in ``tpf_flux``.
+
+        Parameters
+        ----------
+        tpf_flux : array-like
+            A pixel flux time-series, e.g, KeplerTargetPixelFile.flux.
+            Such that (time, row, column) represents the shape of ``tpf_flux``.
+        x0 : array-like or None
+            Initial guesses on the parameters.
+        cadences : array-like of ints or str
+            A list or array that contains the cadences which will be fitted.
+            Default is to fit all cadences.
+        kwargs : dict
+            Dictionary of additional parameters to be passed to
+            `scipy.optimize.minimize`.
+        """
+        if x0 is None:
+            x0 = self.prior.mean
+
+        if cadences == 'all':
+            cadences = range(tpf_flux.shape[0])
+
+        for t in tqdm.tqdm(cadences):
+            loss = self.loss_function(tpf_flux[t], self.scene_model,
+                                      prior=self.prior, **self.loss_kwargs)
+            result = loss.fit(x0=x0, method='powell', **kwargs)
+            opt_params = result.x
+            residuals = tpf_flux[t] - self.scene_model(*opt_params)
+            self.loss_value = np.append(self.loss_value, result.fun)
+            self.opt_params = np.append(self.opt_params, opt_params)
+            self.residuals = np.append(self.residuals, residuals)
+        self.opt_params = self.opt_params.reshape((tpf_flux.shape[0], len(x0)))
+        self.residuals = self.residuals.reshape(tpf_flux.shape)
```

Comment:
I think this function should return `self.opt_params` and explain the return value in the docstring?